chainer/function_hooks/cupy_memory_profile.py

```diff
@@ -0,0 +1,165 @@
+        self._postprocess(function)
+
+    def backward_postprocess(self, function, in_data, out_grad):
+        self._postprocess(function)
+
+    def total_used_bytes(self):
+        """Returns total bytes that functions used from cupy memory pool."""
+        return self._total_used_bytes
+
+    def total_acquired_bytes(self):
+        """Returns total bytes that cupy memory pool acquired from GPU."""
+        return self._total_acquired_bytes
+
+    def summary(self):
+        """Returns a summary of memory profiling in functions.
+
+        Returns:
+            A summarized dictionary whose keys are function names and
+            values are dictionaries of
+            ``used_bytes``, ``acquired_bytes``, and ``occurrrence``.
+        """
+        # TODO(sonots): PROBLEM: takes count of nested functions duplicately
+        summary = {}
+        for func, used_bytes, acquired_bytes, depth in self.call_history:
+            function_name = func.__class__.__name__
+            if function_name not in summary:
+                summary[function_name] = {'used_bytes': 0,
+                                          'acquired_bytes': 0, 'occurrence': 0}
+            record = summary[function_name]
+            record['used_bytes'] += used_bytes
+            record['acquired_bytes'] += acquired_bytes
+            record['occurrence'] += 1
+        return summary
+
+    def _humanized_size(self, size):
+        """Returns a human redable bytes string."""
+        for unit in ['', 'K', 'M', 'G', 'T', 'P', 'E']:
+            if size < 1024.0:
+                return '%3.2f%sB' % (size, unit)
+            size /= 1024.0
+        return '%.2f%sB' % (size, 'Z')
+
+    def print_report(self, file=sys.stdout):
+        """Prints a summary report of memory profiling in functions."""
+        entries = [[
+            'FunctionName', 'UsedBytes', 'AcquiredBytes', 'Occurrence']]
+        for function_name, record in self.summary().items():
+            used_bytes = self._humanized_size(record['used_bytes'])
+            acquired_bytes = self._humanized_size(record['acquired_bytes'])
+            occurrence = str(record['occurrence'])
+            entries.append(
+                [function_name, used_bytes, acquired_bytes, occurrence])
+        entry_widths = []
+        entry_widths.append(max(len(f) for f, _, _, _ in entries))
+        entry_widths.append(max(len(u) for _, u, _, _ in entries))
+        entry_widths.append(max(len(a) for _, _, a, _ in entries))
+        entry_widths.append(max(len(o) for _, _, _, o in entries))
+        template = '  '.join('{:>%d}' % w for w in entry_widths)
+        for function_name, used_bytes, acquired_bytes, occurrence in entries:
+            line = template.format(
+                function_name, used_bytes, acquired_bytes, occurrence)
+            file.write(line)
+            file.write('\n')
+        file.flush()
+
+
+class CupyMemoryCumulativeHook(memory_hook.MemoryHook):
```

Comment:
I want to place at here because moving to cupy has some disadvantages.

* When requirements of cupy memory profile is changed, we have to modify both cupy and chainer
    1. PR will be seperated
    2. CI will fail until cupy side is merged
    3. Users have to upgrade both chainer and cupy